### PR TITLE
[LoRA Attn Processors] Refactor LoRA Attn Processors

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR deprecates all the "LoRA..." attention processors as explained here: https://github.com/huggingface/diffusers/pull/4473#issuecomment-1668269132

This is due to the following reasons:
- Now that we have the concept of `LoRACompatibleLinear` there is no reason to not also leverage it in the attention classes. This way we have a single point of logic, removing overhead.
- It significantly helps when applying global LoRA functionality such as `fuse_lora` as seen here: 
https://github.com/huggingface/diffusers/pull/4473
- We reduce the number of attention processor classes significantly making the file easier to read

TODO:

- [x] Refactor `attn_processor` method to (for now) return derprecated attn processor if LoRA layers are activated
- [x] Make sure the saving format stays the same
- [x] Make sure all LoRA training examples work
- [x] Fix failing test

The PR works now for inference:

```py
pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)
pipe.load_lora_weights("stabilityai/stable-diffusion-xl-base-1.0", weight_name="sd_xl_offset_example-lora_1.0.safetensors")

pipe.to(torch_dtype=torch.float16)
pipe.to("cuda")

torch.manual_seed(0)

prompt = "beautiful scenery nature glass bottle landscape, , purple galaxy bottle"
negative_prompt = "text, watermark"

image = pipe(prompt, negative_prompt=negative_prompt, num_inference_steps=25).images[0]
```

@sayakpaul @williamberman I'd be very happy if you could give this a first review and if ok for you I can finish the final TODOs tomorrow.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Necessary for https://github.com/huggingface/diffusers/pull/4473

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:424]: This function converts the "lora attention processor" key format to the "`LoRACompatibleLinear`" format. After that we can always use the `LoRACompatibleLinear` loading method which allows us to remove a significant amount of code above.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:334]: Would prefer to keep the check and the warning unless we have a solid reason not to.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:351]: Does `convert_state_dict_from_old_format()` always ensure that "lora.down.weight" is always present in the `value_dict`?

File: src/diffusers/loaders.py
Comment (sayakpaul) [reply src/diffusers/loaders.py:424]: We need to account for `LoRACompatibleConv` too.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:437]: Needs to account for `LoRACompatibleConv` too.

File: src/diffusers/loaders.py
Comment (sayakpaul) [reply src/diffusers/loaders.py:424]: I find `old_format` to be slightly misleading. It could mean either the LoRAs when we didn't have text encoder support or the LoRAs that were monkey-patched. Both of them are referred to as the old formats in the codebase. So, maybe "legacy" is better.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:418]: It's not a dict.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:411]: This one corresponds to custom diffusion and seems like an irrelevant change.

File: src/diffusers/models/lora.py
Comment (sayakpaul) [inline src/diffusers/models/lora.py:113]: Good idea to include `lora_scale` in the `forward()`.

**Comment by sayakpaul** [comment]
Did some first pass and left comments. At this point, the changes seem a little too complicated to me but will probably have a better idea once the PR is in a more comprehensive state.

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [reply src/diffusers/loaders.py:411]: Yeah but we need to do this for custom diffusion

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [reply src/diffusers/loaders.py:437]: LoRA attention processors never have a `LoRACompatibleConv` layer, they only have a `LoRACompatibleLinear` layer - so no need to care about `LoRACompatibleConv` here

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [reply src/diffusers/loaders.py:411]: It is not irrelevant in the sense that we now bring it only into the `is_custom_diffusion` clause

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:429]:
```suggestion
            (
                "The decode_latents method is deprecated and will be removed in a future version. Please"
                " use VaeImageProcessor instead"
            ),
```

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:430]:
```suggestion
            (
                "The decode_latents method is deprecated and will be removed in a future version. Please"
                " use VaeImageProcessor instead"
            ),
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py:41]:
```suggestion
        (
            "The preprocess method is deprecated and will be removed in a future version. Please"
            " use VaeImageProcessor.preprocess instead"
        ),
```

**Comment by patrickvonplaten** [comment]
I verified that SDXL LoRA training gives the exact same format when being saved with this PR compared to before. 

@sayakpaul - all tests are now green and I think the PR is more or less ready to go. I would very much like to test it though on some more exotic LoRA formats for inference - I definitely want to test on the following checkpoints:

> https://civitai.com/models/22279?modelVersionId=118556 (UNet attention LoRA)
> https://civitai.com/models/104515/sdxlor30costumesrevue-starlight-saijoclaudine-lora (UNet attention and conv LoRA)
> https://civitai.com/models/108448/daiton-sdxl-test (everything combined LoRA)

from https://github.com/huggingface/diffusers/pull/4287#issue-1822372761 

Are there more checkpoints you would recommend me to test (maybe some non-SDXL ones?).

**Comment by sayakpaul** [comment]
You can find some in the integration tests of `test_lora_layers.py`.

**Comment by patrickvonplaten** [comment]
**A bit more explanation for what's going on here**:

- a) Eventually we want to fully get rid of "lora" logic in `diffusers` and offload all this responsability to PEFT. The only thing that will be left in diffusers then will be `{AnyLoraFormat}->PEFT_Format` conversion logic which probably only amounts to renaming keys and other inputs.
- b) In this PR in a first step we want to unify the LoRA logic all into `LoRACompatibleLinear` and essentially deprecate all Attn Processor LoRAs. After this PR, all the computation logic will flow through `LoRACompatibleLinear` and we keep the LoRA attention processors only as a serialization format to not break backwards compatibility. Hence we need to ensure that calling `unet.attn_processors` will still return LoRA attention processors so that all the serialization format stays intact.
- c) Running the forward pass of any LoRA attention processor is already now deprecated. The LoRA attention processor serialization format will bit by bit be deprecated, but this should be done as part of the PEFT integration.

Overall this PR should make the whole LoRA loading and functioning easier to understand and build upon (such as needed for #4473

File: tests/models/test_lora_layers.py
Comment (patrickvonplaten) [inline tests/models/test_lora_layers.py:926]: Updating the expected values here because I'm quite sure that this PR corrects a bug.
The lora: 

```
        lora_model_id = "hf-internal-testing/sdxl-0.9-kamepan-lora"
        lora_filename = "kame_sdxl_v2-000020-16rank.safetensors"
```

is quite unique since it has different `network_alphas` for q,k,v and out LoRA of the attention layer. Previously we always set the same `network_alpha` for all q,k,v and out (see [here](https://github.com/huggingface/diffusers/blob/3bba44d74eaf20af0d14424d786381f0103c6c90/src/diffusers/models/attention_processor.py#L546))

However, it might very well be that q,k,v and out have different `network_alpha` values. This is now possible as of this PR and therefore should mean it's a bug fix since this checkpoint has different `network_alpha`s for each q,k,v and out.

Comment (sayakpaul) [reply tests/models/test_lora_layers.py:926]: https://github.com/huggingface/diffusers/blob/3bba44d74eaf20af0d14424d786381f0103c6c90/src/diffusers/loaders.py#L1407 suggests we assign different alphas to the modules you mentioned. What am I missing out on?

Comment (sayakpaul) [reply tests/models/test_lora_layers.py:926]: Ah got it. We do that for the text encoder only it seems but not the UNet, is my understanding correct?

Comment (patrickvonplaten) [reply tests/models/test_lora_layers.py:926]: Yeah exactly! I'm pretty sure it's corrected now :-)

**Review by williamberman — APPROVED** [review]
nice nice

**Comment by patrickvonplaten** [comment]
As soon as you're good with the PR @sayakpaul - I think we can merge this one and unblock the fuse_lora PR :-)

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:358]: Maybe we should just display `list(is_lora_activated.keys())` to the end user.

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:311]: Neat. 

I am a little concerned about:

```
and self.to_q.lora_layer is not None
```

i.e., only check if `to_q.lora_layer` is not None. Is there a better alternative?

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:361]: This bit is nasty but needed for ensuring backwards compatibility.

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1359]: Do you want to directly throw a deprecation message when the class is initialized or would that be too brutal?

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1423]: Likewise.

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1488]: Like wise.

File: tests/models/test_lora_layers.py
Comment (sayakpaul) [reply tests/models/test_lora_layers.py:756]: You meant 1e-4.

**Review by sayakpaul — APPROVED** [review]
Hard work. Thanks a lot. 

Left some minor design-related comments but I don't think they block this PR from getting merged. So, I am gonna merge this to be able to focus on https://github.com/huggingface/diffusers/pull/4473.

File: src/diffusers/models/attention_processor.py
Comment (patrickvonplaten) [reply src/diffusers/models/attention_processor.py:1359]: I thought about this too - think this will be a bit too much though at the moment because it would mean every time we call `unet.attn_processors` we would get a deprecation warning which would be every time we save LoRAs. I'd suggest to only start throwing aggressive deprecation warnings when we do the peft integration

**Pull request merged**
